### PR TITLE
Support a simple dns records relation

### PR DESCRIPTION
**Pull request by nrobinaubertin**
### Overview

This adds support for a simple relation in the charm and an integration test for it using any-charm.  
It is very simple for now:
- only one relation
- only one entry request
- no checking of the result

It needs https://github.com/canonical/bind-operator/pull/16 to be merged first.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [ ] The documentation for charmhub is updated
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

The documentation will be updated when the full merging process is implemented.

### CODE REVIEW

**Review by mthaddon — COMMENTED** [review]
Some minor comments but otherwise looking great

File: src/bind.py [hunk 1]
```diff
@@ -114,3 +127,96 @@ def _install_snap_package(
...
+        zones_entries: typing.Dict[str, typing.List[RequirerEntry]] = {}
+        for entry in rrd.dns_entries:
+            if entry.domain not in zones_entries:
+                zones_entries[entry.domain] = []
+            zones_entries[entry.domain].append(entry)
+
+        zones_content: typing.Dict[str, str] = {}
+        for zone, entries in zones_entries.items():
+            content: str = (
+                f"$ORIGIN {zone}.\n"
```
Comment (arturo-seijas) [inline src/bind.py:160]: Perhaps this template should be a constant?

File: src/bind.py [hunk 2]
```diff
@@ -114,3 +127,96 @@ def _install_snap_package(
...
+        Args:
+            zones: A list of all the zones names
+
+        Returns:
+            The content of `named.conf.local`
+        """
+        # It's good practice to include rfc1918
+        content: str = f'include "{constants.DNS_CONFIG_DIR}/zones.rfc1918";\n'
+        for name in zones:
+            content += (
```
Comment (arturo-seijas) [inline src/bind.py:187]: Same here

File: src/bind.py (hunk 1 above)
Comment (nrobinaubertin) [reply src/bind.py:160]: I can't store an f-string for later use as it has to be evaluated at creation. If our needs grow, we could use jinja.

File: tests/integration/test_charm.py [hunk 3]
```diff
@@ -93,3 +95,48 @@ async def test_basic_dns_config(app: ops.model.Application, ops_test: OpsTest):
...
+        config={
+            "src-overwrite": json.dumps(any_charm_src_overwrite),
+            "python-packages": "pydantic==2.7.1\n",
+        },
+    )
+    await ops_test.model.wait_for_idle(status="active")
+
+    await ops_test.model.add_relation(f"{any_charm.name}", f"{app.name}")
+
+    # give some time to the bind charm to handle the request
```
[the remark this answers is not in the document]
Comment (nrobinaubertin) [reply tests/integration/test_charm.py]: It works locally based on the status. Changed

File: src/bind.py [hunk 4]
```diff
@@ -114,3 +127,96 @@ def _install_snap_package(
             error_msg = f"An exception occurred when installing {snap_name}. Reason: {e}"
             logger.error(error_msg)
             raise InstallError(error_msg) from e
+
+    def _push(self, path: pathlib.Path, source: str) -> None:
```
[the remark this answers is not in the document]
Comment (nrobinaubertin) [reply src/bind.py]: Renamed

File: src/bind.py (hunk 1 above)
Comment (arturo-seijas) [reply src/bind.py:160]: You can define the string as a constant and then evaluate it later

Comment (nrobinaubertin) [reply src/bind.py]: Understood. I used `str.format()` as a lightweight templating system

**Review by arturo-seijas — APPROVED** [review]

File: src/bind.py (hunk 2 above)
Comment (nrobinaubertin) [reply src/bind.py]: Same resolution

**Review by mthaddon — APPROVED** [review]

**Pull request merged**